lib/Conversion/BGVToOpenfhe/BGVToOpenfhe.cpp

```diff
@@ -0,0 +1,264 @@
+      : OpConversionPattern<Rotate>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      Rotate op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    FailureOr<Value> result = getContextualCryptoContext(op.getOperation());
+    if (failed(result)) return result;
+
+    Value cryptoContext = result.value();
+    auto offsetValue = rewriter.create<arith::ConstantOp>(
+        op.getLoc(), rewriter.getIntegerAttr(rewriter.getIntegerType(64),
+                                             adaptor.getOffset()));
+    rewriter.replaceOp(
+        op, rewriter.create<openfhe::RotOp>(op.getLoc(), cryptoContext,
+                                            adaptor.getX(), offsetValue));
+    return success();
+  }
+};
+
+bool checkRelinToBasis(llvm::ArrayRef<int> toBasis) {
+  if (toBasis.size() != 2) return false;
+  return toBasis[0] == 0 && toBasis[1] == 1;
+}
+struct ConvertRelinOp : public OpConversionPattern<Relinearize> {
+  ConvertRelinOp(mlir::MLIRContext *context)
+      : OpConversionPattern<Relinearize>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      Relinearize op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    FailureOr<Value> result = getContextualCryptoContext(op.getOperation());
+    if (failed(result)) return result;
+
+    auto toBasis = adaptor.getToBasis();
+    if (!checkRelinToBasis(toBasis)) {
+      op.emitError() << "toBasis must be [0, 1], got "
+                     << "[" << toBasis << "]";
+      return failure();
+    }
+
+    Value cryptoContext = result.value();
+    rewriter.replaceOp(op, rewriter.create<openfhe::RelinOp>(
+                               op.getLoc(), cryptoContext, adaptor.getX()));
+    return success();
+  }
+};
+
+bool checkModulusSwitchLevels(unsigned long fromLevel, unsigned long toLevel) {
+  return fromLevel == toLevel + 1;
+}
+struct ConvertModulusSwitchOp : public OpConversionPattern<ModulusSwitch> {
+  ConvertModulusSwitchOp(mlir::MLIRContext *context)
+      : OpConversionPattern<ModulusSwitch>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      ModulusSwitch op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    FailureOr<Value> result = getContextualCryptoContext(op.getOperation());
+    if (failed(result)) return result;
+
+    auto fromLevel = adaptor.getFromLevel();
+    auto toLevel = adaptor.getToLevel();
+    if (!checkModulusSwitchLevels(fromLevel, toLevel)) {
+      op.emitError() << "fromLevel must be toLevel + 1, got "
+                     << "fromLevel: " << fromLevel << " and "
+                     << "toLevel: " << toLevel;
+      return failure();
+    }
+
+    Value cryptoContext = result.value();
+    rewriter.replaceOp(op, rewriter.create<openfhe::RelinOp>(
```

Comment:
```suggestion
    rewriter.replaceOp(op, rewriter.create<openfhe::ModReduceOp>(
```

(or perhaps it's LevelReduce? I'm not sure the difference)